### PR TITLE
Fix AutocompleteInput `clearOnBlur` has no effect

### LINKED ISSUE #8096: Improper input value selection process for Autocomplete/ReferenceInput combo
**What you were expecting:**
Consistent behavior from AutocompleteInputs as children of ReferenceInputs.

**What happened instead:**
The AutocompleteInput will save incomplete text inputs without having made a dropdown selection. 

The ReferenceInput does not seem to be making the appropriate calls when given the prop "enableGetChoices"

**Steps to reproduce:**
Add an AutocompleteInput with common props such as filterToQuery and shouldRenderSuggestions as a child of a ReferenceInput with common props such as enableGetChoices.

**Related code:**
https://codesandbox.io/s/ecstatic-ptolemy-64jbky?file=/src/posts/PostCreate.tsx

**Environment**

* React-admin version: 4.3.0
* Last version that did not exhibit the issue (if applicable): 4.2.7
* React version: 17
* Browser: Chrome

### PR DESCRIPTION
**Pull request by WiXSL**
Fixes #8096

### CODE REVIEW

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 1]
```diff
@@ -213,6 +214,13 @@ export const AutocompleteInput = <
         ...rest,
     });
 
+    const finalOnBlur = useCallback((): void => {
+        if (clearOnBlur && !(!!create || !!onCreate)) {
+            setFilterValue('');
+        }
```
Comment (slax57) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:220]: I don't understand, why clear the filter only if you don't have `create` or `onCreate` props?

Comment (WiXSL) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:220]: Is our custom `freeSolo` condition ( `freeSolo={!!create || !!onCreate}`).
At first I was asking `if (clearOnBlur && !props.freeSolo) {`

Comment (slax57) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:220]: Not sure I understand :grimacing: 
According to https://mui.com/material-ui/react-autocomplete/#creatable , we should have `clearOnBlur=true` when `freeSolo=true`.

Since having `clearOnBlur=true` on MUI's `<Autocomplete>` does not seem to work, to me the condition should be `if (clearOnBlur || props.freeSolo)` rather than `if (clearOnBlur && !props.freeSolo)`
Or simply `if (clearOnBlur)` since we encourage `clearOnBlur=true` by default anyway.

Again, not sure I understand your point completely, but I tested your branch with the simple example, and it feels weird to still have something inside the Autocomplete when I click away and did not create a new item.

TLDR; I'd simply go with `if (clearOnBlur)`

**Review by slax57 — APPROVED** [review]
Thanks :slightly_smiling_face: 

Now I just would like to have @fzaninotto 's opinion on this before merging, just to be sure

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 2]
```diff
@@ -213,6 +214,13 @@ export const AutocompleteInput = <
         ...rest,
     });
 
+    const finalOnBlur = useCallback((): void => {
+        if (clearOnBlur) {
+            setFilterValue('');
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:219]: I've tested it with the storybook, and it feels weird. Because if I blur the field without selecting anything, the field appears empty *but the actual value hasn't changed*, so the SaveButton isn't enabled. 

We should set the filter value to the current value in this case.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 3]
```diff
@@ -233,6 +226,14 @@ export const AutocompleteInput = <
         optionValue,
     });
 
+    const finalOnBlur = useCallback((): void => {
+        if (clearOnBlur) {
+            const optionLabel = getOptionLabel(selectedChoice);
+            setFilterValue(optionLabel);
+        }
+        field.onBlur();
+    }, [clearOnBlur, field, selectedChoice]);
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:235]: linter warning

**Review by fzaninotto — APPROVED** [review]
